### PR TITLE
Feat/dispatcher user crud

### PR DESCRIPTION
**Pull request by Parle648**
## Describe your changes

- Create dispatcher Get one dispatcher Update dispatcher Delete dispatcher.
 - Available only for admin from the same company

## Issue ticket code (and/or) and link

- [Link to JIRA ticket](#https://ticket-url)

### **General**

- [x] types for input and output params
- [x] no `any`, should be also added to eslint rules
- [x] try/catch for any async function
- [x] no **magic** [numbers](/7d77574ee4bc4e339a68066762e887cb)
- [x] compare only with constants not with strings (instead of user === ‘admin’, better user === roles.admin)
- [x] no ternary operator inside ternary operator (bad example: question ? first: second ? third: fourth)
- [x] avoid similar types with duplicating by generics
- [x] no **console.log** in pr
- [x] .env file should be in .gitignore
- [x] delete commented code if it's not part of documentation
- [x] no links in the code, env links should be in env file (for example: server url), constant links (for example default avatar URL) should be in constant file
- [x] **constants** and **function** names should be lowercase.
- [x] no dates format in the code (like "dd/MM/yyyy”), move it in global constant variable
- [x] import rules. imports should come in a specific order: node modules, absolute path, relative path
- [x] strings should be in the constant variable. Example: instead of ‘15 3 \* \* _’, const EACH_DAY_15h_15min = ‘15 3 _ \* \*’

### Backend

- [x] swagger for each endpoint. add in the documentation, different types of responses.
      on a successful response (2xx), on unsuccessful response (4xx, 5xx)
- [x] less requests to db, all data should be taken with one query
- [x] **public** in method use only is use function externally
- [x] use ConfigService instead of **process.env**
- [x] use @`@index` decorator for frequently requested data
- [x] use transactions if there is a call chain that mutates data in different tables
- [x] add a set of rules for [nestjs](https://github.com/darraghoriordan/eslint-plugin-nestjs-typed) to the .eslint config file. It helps to prevent bugs and confusing moments
- [x] use [REST API Naming Convention](https://medium.com/@nadinCodeHat/rest-api-naming-conventions-and-best-practices-1c4e781eb6a5) for endpoints

  P.S: more information about RESTful API design by [Microsof](https://learn.microsoft.com/en-us/azure/architecture/best-practices/api-design)t

- [x] use ‘**UUIDs**` for primary key, ** to prevent Enumeration Attacks**

### CODE REVIEW

File: src/modules/user/dispatcher/dispatcher.service.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+export class DispatcherService {
+  constructor(
+    @InjectRepository(User)
+    private readonly userRepo: Repository<User>,
+    @InjectRepository(Company)
+    private readonly companyRepo: Repository<Company>,
+    private readonly entityManager: EntityManager,
+  ) {}
+
+  async create(createDispatcherDto: CreateDispatcherDto, authorization: string): Promise<User> {
```
Comment (lyudad) [inline src/modules/user/dispatcher/dispatcher.service.ts]: nono logic is not right, service should not work with the token, during validation, you can return the user object to route, and from route to service, but not a token
in guard get user => controller => service

File: src/modules/user/dispatcher/dispatcher.service.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+      if (company.id !== decodedToken.company_id.id) throw new BadRequestException("You doesn't work in this company");
+
+      const dispatcher = this.userRepo.create({ ...createDispatcherDto, company_id: company });
+      return await this.entityManager.save(dispatcher);
+    } catch (error) {
+      throw new InternalServerErrorException('Internal server error');
+    }
+  }
+
+  async findOne(id: number, authorization: string): Promise<User> {
```
Comment (lyudad) [inline src/modules/user/dispatcher/dispatcher.service.ts]: same

**Review by lyudad — APPROVED** [review]

File: src/modules/user/dispatcher/dispatcher.service.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+      const decodedToken = <AccessTokenInterface>jwt.decode(authorization);
+      if (dispatcher.company_id.id !== decodedToken.company_id.id) throw new BadRequestException("You doesn't work in this company");
+      return await this.userRepo.findOneByOrFail({ id });
+    } catch (error) {
+      throw new BadRequestException('There is no such dispatcher');
+    }
+  }
+
+  async update(id: number, updateDispatcherDto: UpdateDispatcherDto, authorization: string): Promise<ResponseInterface> {
```
Comment (lyudad) [inline src/modules/user/dispatcher/dispatcher.service.ts]: same

File: src/modules/user/dispatcher/dispatcher.service.ts (hunk 1 above)
Comment (Parle648) [reply src/modules/user/dispatcher/dispatcher.service.ts]: fix

File: src/modules/user/dispatcher/dispatcher.service.ts (hunk 2 above)
Comment (Parle648) [reply src/modules/user/dispatcher/dispatcher.service.ts]: fix

File: src/modules/user/dispatcher/dispatcher.service.ts (hunk 3 above)
Comment (Parle648) [reply src/modules/user/dispatcher/dispatcher.service.ts]: fix

**Review by lyudad — APPROVED** [review]

**Pull request merged**